### PR TITLE
Page table cache

### PR DESCRIPTION
**Pull request by jenswi-linaro**

### CODE REVIEW

File: architecture/core.rst [hunk 1]
```diff
@@ -385,6 +385,56 @@ context is activated.
...
     }
 
+Page table cache
+================
+Page tables used to map TAs are managed with the page table cache. When the
+context of a TA is unmapped, all its page tables are released with a call
+to ``pgt_free()``. All page tables needed when mapping a TA are allocated
+using ``pgt_alloc()``.
+
+A fixed maximum number of translation tables are available in a pool.  One
```
Comment (jbech-linaro) [inline architecture/core.rst]: nit: Double spaces (`.  One`)

File: architecture/core.rst [hunk 2]
```diff
@@ -385,6 +385,56 @@ context is activated.
...
+thread may be executing a TA which needs all or almost all tables. This can
```
Comment (jbech-linaro) [inline architecture/core.rst]: s/may be executing/may execute/

File: architecture/core.rst [hunk 3]
```diff
@@ -385,6 +385,56 @@ context is activated.
...
+block TAs from being executed by other threads. To make sure that all TAs
```
Comment (jbech-linaro) [inline architecture/core.rst]: s/make sure/ensure/

File: architecture/core.rst [hunk 4]
```diff
@@ -385,6 +385,56 @@ context is activated.
...
+format (``CFG_WITH_LPAE=n``) it becomes more complicated as four
+translation tables are stored in each page. Additional bookkeeping is used
+to tell when the page for used by four separate translation tables can be
+released.
+
+**With paging of user TA enabled,** ``CFG_PAGED_USER_TA=y``
+
+With paging of user TAs enabled a cache of recently used translation tables
+is used. This can save us from a storm of page faults when restoring the
+mappings of a recently unmapped TA. Which transaltion tables should be
```
Comment (jbech-linaro) [inline architecture/core.rst]: s/transaltion/translation/

File: architecture/core.rst [hunk 5]
```diff
@@ -385,6 +385,56 @@ context is activated.
...
+cached is indicated with refernce counting by the pager on used tables.
```
Comment (jbech-linaro) [inline architecture/core.rst]: s/refernce/reference/

File: architecture/core.rst [hunk 6]
```diff
@@ -385,6 +385,56 @@ context is activated.
...
+When a table needs to be forcfully freed
```
Comment (jbech-linaro) [inline architecture/core.rst]: s/forcfully/forcefully/

File: architecture/core.rst [hunk 7]
```diff
@@ -385,6 +385,56 @@ context is activated.
...
+A fixed maximum number of translation tables are available in a pool.  One
+thread may be executing a TA which needs all or almost all tables. This can
+block TAs from being executed by other threads. To make sure that all TAs
+eventually will be permitted to execute ``pgt_alloc()`` temporarily frees
+eventual tables allocated before waiting for tables to become available.
+
+The page table cache behaves differently depending on configuration
+options.
+
+**Without paging,** ``CFG_WITH_PAGER=n``
```
Comment (jbech-linaro) [inline architecture/core.rst]: I would probably do like this (makes it linkable also and you can remove the empty line ... 405).
```rst
Without paging (``CFG_WITH_PAGER=n``)                                                                                                                                                                              
-------------------------------------
```

File: architecture/core.rst [hunk 8]
```diff
@@ -385,6 +385,56 @@ context is activated.
...
+
+This is the easiest configuration. All page tables are statically allocated
+in the ``.nozi.pgt_cache`` section. ``pgt_alloc()`` allocates tables from the
+free-list and ``pgt_free()`` returns the tables directly to the free-list.
+
+**With paging enabled,** ``CFG_WITH_PAGER=y``
```
Comment (jbech-linaro) [inline architecture/core.rst]: And just as above, like this:
```rst
With paging enabled (``CFG_WITH_PAGER=y``)                                                                                                                                                                         
------------------------------------------
```

File: architecture/core.rst [hunk 9]
```diff
@@ -385,6 +385,56 @@ context is activated.
...
+
+With ``CFG_WITH_LPAE=y`` each translation table has the same size as a
+physical page which makes it easy to release the physical page when the
+translation table isn't needed any longer. With the short-descriptor table
+format (``CFG_WITH_LPAE=n``) it becomes more complicated as four
+translation tables are stored in each page. Additional bookkeeping is used
+to tell when the page for used by four separate translation tables can be
+released.
+
+**With paging of user TA enabled,** ``CFG_PAGED_USER_TA=y``
```
Comment (jbech-linaro) [inline architecture/core.rst]:
```rst
With paging of user TA enabled (``CFG_PAGED_USER_TA=y``)                                                                                                                                                           
--------------------------------------------------------
```

File: architecture/core.rst (hunk 7 above)
Comment (jenswi-linaro) [reply architecture/core.rst]: OK, I'll fix. The reason I didn't do it like this from the beginning was that I though it was a bit hard to see that it really was a subsection to "Page table cache"

**Comment by jenswi-linaro** [comment]
Comments addressed

**Comment by jbech-linaro** [comment]
`Reviewed-by: Joakim Bech <joakim.bech@linaro.org>`

**Comment by jenswi-linaro** [comment]
Tag applied

**Pull request merged**